### ISSUE TITLE
Box integration

### ISSUE THREAD

**Issue by theofidry** [humbug/php-scoper#49]
As people use [box](https://github.com/humbug/box) to build their PHAR, it would be nice if the prefixing could be done directly via box by exposing a config.

This is very likely to require to switch from a JSON config file to a PHP one for Box.

**Comment by theofidry** [comment]
To be tracked on #31 now :)

**Comment by TomasVotruba** [comment]
I wonder if these 2 tools could be bundled together as `phar-builder`, just with separate 2 commands.
They share so much intell and are pretty useless one without another, moreover with 1 maintainer it would save so much redundant work.

Just an idea :+1:

**Comment by theofidry** [comment]
Oups, look like I linked the wrong issue, or rather the right number but the wrong repo. I meant https://github.com/humbug/box/pull/31.

I would say partly yes and partly no. I feel like PHP-Scoper is big & complex enough to be a tool on its own. It can easily be used outside of Box and even outside of the context of PHARs. For example for a tool like SensioLabs Insight, sure you can use an isolated PHAR, but what you really need is the isolated part. It being a PHAR would be to make it more convenient to distribute or something. Another example would be if you are not using Box but another tool to build your PHARs.

But in any case I 100% agree with you, Box _needs_ to provide a PHP-Scoper integration. Which is partly why I didn't tag Box yet. I'm waiting for https://github.com/humbug/box/pull/31 to be completed to tag the first alpha release

**Comment by TomasVotruba** [comment]
```bash
phar-builder build
phar-builder prefix
```

**Comment by theofidry** [comment]
Why would you prefer to have two separate commands rather than just having to do `box build`?

**Comment by TomasVotruba** [comment]
(I've updated my comment)

**Comment by TomasVotruba** [comment]
My big inspiration is https://github.com/fprochazka/phpstan-compiler/tree/master/src/Compiler which work in 3 classes for [PHPStan-shim](https://github.com/phpstan/phpstan-shim).

**Comment by theofidry** [comment]
Sure, but it's also way more opinionated.

I'm with you that those two projects look overly complex. There's the feeling of the complexity in developing/using it and there's also the fact that there is a lot of alternative ways to do it that are very simple.

So to recap, when I took box over a few months ago, Box was depending on no less than 6 other packages which were created just for Box (all marked as deprecated). Feature-wise it was more complete as well. To me that just looked insane and I started to refactor it: there's only one repository for Box no and it is _a lot_ simpler. For the record about 15KLoC was added when merging together, since then about 10K have been removed although the total is higher thanks to adding more tests.

So complexity wise, it is getting better. Now why is it still complex? Configuration & Architecture. The configuration is more extensible and less opinionated in Box, if you look at [the Box Configuration class](https://github.com/humbug/box/blob/master/src/Configuration.php) it's the biggest class of the project by far. I'm working to simplify it, but the simplifications are likely to be for the users not really code-wise. The only way to massively reduce it code-wise is to be more opinionated. For example like in PHPStan-shim: force the user to have more of his own config to do instead of doing the hard work in the tool or Laravel-Zero which makes a lot of assumptions on how the application is looking.

Architecture-wise for the current usage _it is_ overkill. It could be greatly simplified IMO (and I already simplified it a lot). But still want to keep the base design because it makes more sense to solve the problem I'm trying to tackle (which the other tools won't help with).

Regarding the prefixing:
- PHP-Scoper is also meant to be used outside of the context of PHARs: for regular applications, Wordpress plugins or even with other PHAR builders. As such requiring a PHAR builder application for it looks weird. I'm not ruling out to merge the two projects as it would indeed make it easier for me and I'm a big fan of mono-repos, but I think it's too early for that.
- The PHPStand compiler _is_ interesting, but I experimented with it (actually the first approach of PHP-Scoper was based on it), but you encounter too many weird cases for it to be viable on the long-run. The only reason why it's working ok with PHPStan is because PHPStan codebase is ok (there is very few dynamic stuff).

TL:DR; Yes there is simpler alternatives, but they won't help to tackle the issues I'm trying to tackle. That said I do aim for things to be simple for the user (way more than what you're currently experiencing).

I'm still not clear however on why you would (as a user) prefer to do:

```
phar-builder build
phar-builder prefix
```

as opposed to:

```
phar-builder build # handle prefixed on not in the config
```

**Comment by TomasVotruba** [comment]
> as opposed to:
>
> > phar-builder build # handle prefixed on not in the config

Even simpler better! 

My main point is why develop 2 packages, that need to work with each other, in 1 vendor but in 2 repositories?
Why not [Go monorepo](https://gomonorepo.org/)?

**Comment by theofidry** [comment]
>My main point is why develop 2 packages, that need to work with each other, in 1 vendor but in 2 repositories?

History. I started PHP-Scoper (which as I said don't need Box at all) first and _then_ forked Box. Note that the namespaces in Box didn't change yet and that's because maybe eventually this can be merged back to the original project. That will depend on how it grows.

But you're right, one thing that could be done is having PHP-Scoper in Box and expose PHP-Scoper as a standalone tool as well. That could be done but it would need some adjustments plus a proper migration. But right now I'm very short on time and I really would like to get this alpha out before thinking of this kind of stuff :)

>Why not Go monorepo?

Well except the example of PHP-Scoper and Box, about 90% of my projects are on orga I either don't have full control on, wouldn't make sense to merge or simply it would break stuff for too many users. Example: nelmio, hautelook, infection, myclabs, wouterj, api-platform... Humbug is the real exception here :P
